drv/lpc55-swd/src/main.rs

```diff
@@ -484,16 +489,52 @@ impl idl::InOrderSpCtrlImpl for ServerImpl {
             Err(_) => Err(SpCtrlError::Fault.into()),
         }
     }
+
+    fn enable_sp_slot_watchdog(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+        time_ms: u32,
+    ) -> Result<(), RequestError<SpCtrlError>> {
+        ringbuf_entry!(Trace::EnabledWatchdog);
+        if !self.init {
+            return Err(SpCtrlError::NeedInit.into());
+        }
+        // This function is idempotent(ish), so we don't care if the timer was
+        // already running; set the new deadline based on current time.
+        let deadline = sys_get_timer().now + time_ms as u64;
+        sys_set_timer(Some(deadline), notifications::TIMER_MASK);
+        Ok(())
+    }
+
+    fn disable_sp_slot_watchdog(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+    ) -> Result<(), RequestError<core::convert::Infallible>> {
+        ringbuf_entry!(Trace::DisabledWatchdog);
+        sys_set_timer(None, notifications::TIMER_MASK);
+        Ok(())
+    }
 }
 
 impl NotificationHandler for ServerImpl {
     fn current_notification_mask(&self) -> u32 {
-        // We don't use notifications, don't listen for any.
-        0
+        notifications::TIMER_MASK
     }
 
     fn handle_notification(&mut self, _bits: u32) {
-        unreachable!()
+        ringbuf_entry!(Trace::WatchdogFired);
+
+        // Disable the watchdog timer
+        sys_set_timer(None, notifications::TIMER_MASK);
+
+        // Attempt to do the swap
+        match self.swap_sp_slot() {
+            Ok(()) => ringbuf_entry!(Trace::WatchdogSwapOk),
+            Err(e) => ringbuf_entry!(Trace::WatchdogSwapErr(e)),
+        }
```

Comment:
nit, take it or leave it: it _could_ be a little bit simpler to just have a `Trace::WatchdogSwap(Result<(), Ack>)` and then write
```suggestion
        // Attempt to do the swap
        let result = self.swap_sp_slot();
        ringbuf_entry!(Trace::WatchdogSwap(result));
```
(and, if we wanted to add counters to this ringbuf later, we can add `#[count(children)]` as `Result<(), E> where E: Count` implements `Count`)...

Reply:
Sure, done in b990e10bf929227aeb88490aa3a76c6b029f3c68